docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,146 @@
+* The various jobs can communicate by storing temporary data in the caching service or in the Blobstore. If the latter is used, the FI job should clean up the Blobstore nodes when its complete.
+* Could make a helper app for this?
+* What about workflow engines (WDL, Cromwell)? Are we reinventing the wheel here?
+* Can new EE2 endpoints speed up or reduce the complexity of any of these steps?
+
+### Notes about DAG in ee2 Endpoints
+```
+Your dag would need to have (at least) a first job followed by a SUBDAG EXTERNAL.
+Somewhere in the first job you'd generate a new dag workflow that
+defines the N clusters followed by the N+1 job, which runs in the
+subdag.
+
+As for DAGMan support in the Python bindings, we do this in the
+following two ways:
+
+1) There is a htcondor.Submit.from_dag() option which takes the name
+of a dag filename. You then submit the resulting object just like any
+regular job.
+2) We have a htcondor.dags library which can be used to
+programmatically construct a DAG workflow in computer memory, then
+write to a .dag file and submit using the function mentioned in 1)
+above.
+```
+
+Between these there are several different ways to do what you want.
+
+There's a useful example here that shows the general workflow in the
+bindings: https://htcondor.readthedocs.io/en/latest/apis/python-bindings/tutorials/DAG-Creation-And-Submission.html#Describing-the-DAG-using-htcondor.dags
+
+## Consequences
+
+* We will have to roll out fixes in multiple stages
+* We will have to implement a new narrative UI, however this was work that would happen regardless due as we are looking to improve the UX for batch upload and analysis at KBase. 
+* This will take significant time to further research and engineer the solutions
+
+Still to be determined (not in scope of this ADR): 
+* UI and how it relates to the bulk execution
+* XSV Analysis and how it relates to the bulk execution
+
+## Alternatives Considered
+
+* Ignore most issues and just make apps that run kbparallels limited to N instances of kbparallels per user to avoid deadlocks
+* Writing new ee2 endpoints to entirely handle batch execution and possibly use a DAG
+* Remove kbparallels and change apps to a collection of 2-3 apps that do submit, split and aggregate and an use an ee2 endpoint to create a DAG
+* Different DevOps solutions
+
+## Pros and Cons of the Alternatives
+
+### Limit multiple instances of kbparallels
+
+* `+` Simplest solution, quickest turnaround, fixes deadlock issue
+* `-` Addresses only the deadlocking issue, UI still broken for regular runs and batch runs 
+
+### Increase number of slots or Seperate Queue for kbparallels apps without 10 job limit
+* `+` Simple solutions, quick turnarounds, fixes deadlock issue
+* `-` Addresses only the deadlocking issue, UI still broken for regular runs and batch runs
+* `-` A small amount of users can take over the entire system
+* `-` The calculations done by the apps will interfere with other apps and cause crashes/failures
```

Comment:
I think these two are true of we increase the number of slots, but not if KBP apps have a separate queue. Maybe those two solutions should be separate with individual pro/con lists?

Reply:
@MrCreosote Which two?

Reply:
changed